Zip/Functions/Add-ZipArchiveEntry.ps1

```diff
@@ -0,0 +1,168 @@
+
+    .EXAMPLE
+    Get-ChildItem 'C:\Projects\Zip' -File | Add-ZipArchiveEntry -ZipArchivePath 'zip.zip' -BasePath 'C:\Projects\Zip'
+
+    Demonstrates how to control the paths of files in the ZIP archive. In this case, files will *not* start with `Projects\Zip`. `Add-ZipArchiveEntry` removes the `BasePath` from the start of each file's path when determining its path in the ZIP file.
+
+    .EXAMPLE
+    Add-ZipArchiveEntry -ZipArchivePath 'zip.zip' -InputObject 'Zip' -BasePath (Get-Location).Path
+
+    Demonstrates how to pass a path to add that directory/file to a ZIP file. In this example, the contents of the `Zip` directory in the current directory will be added to the ZIP file. Because the `BasePath` parameter is used, the files in the file will begin with `Zip` instead of the full path to the Zip directory.
+
+    .EXAMPLE
+    Get-Item -Path '.\Zip' | Add-ZipArchiveEntry -ZipArchivePath 'zip.zip' -BasePath (Get-Location).Path -EntryParentPath 'package'
+
+    Demonstrates how to customize the directory in the ZIP file files will be added at. In this case, all the files under the `Zip` directory will be put in a `packages` directory, e.g. `packages\Zip`.
+
+    .EXAMPLE
+    Get-ChildItem 'C:\Projects\Zip' | Add-ZipArchiveEntry -ZipArchivePath 'zip.zip' -BasePath 'C:\Projects\Zip' -EntryParentPath 'Zip2'
+
+    Demonstrates how to give a directory a different name when adding it to the zip file. In this case, we're adding all the items in `C:\Projects\Zip`. Because the `BasePath` parameter is also set to `C:\Projects\Zip`, normally all the files/directories in `C:\Projects\Zip` would be in the root of the ZIP file, but because the `EntryParentPath` parameter is set to `Zip2`, all the files/directories will be put in a `Zip2` directory.
+    #>
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory)]
+        [string]
+        # The path to the ZIP file. Files will be added to this ZIP archive.
+        $ZipArchivePath,
+
+        [Parameter(Mandatory,ValueFromPipeline,ValueFromPipelineByPropertyName)]
+        [Alias('FullName')]
+        [Alias('Path')]
+        [string]
+        # The files/directories to add to the archive. Normally, you would pipe file/directory objects to `Add-ZipArchiveEntry`. You may also pass any object that has a `FullName` or `Path property. You may also pass the path as a string.
+        #
+        # If you pass a directory object or path to a directory, all files in that directory and all its sub-directories will be added to the archive.
+        $InputObject,
+
+        [string]
+        # By default, items are added to the ZIP archive using their full paths (minus any drives/qualifiers). The `BasePath` parameter controls what portion of the source file's path is removed when determining its name in the ZIP archive.

```

Comment:
This strikes me as odd that this is the default behavior. As a user, I've been accustomed to relative paths always being the default behavior for commands. I feel this should be the opposite, if a user specifies `C:\projects\zip` then the path in the ZIP archive should be simply `zip`, not `projects\zip`. This is how archive tools such as 7-zip and the Zip utility built into Windows work by default.

What if we add a `PathMode` parameter, similar to how 7-zip employs it? For example:

Given path: `C:\projects\zip`

Path Mode: `Relative`
```
zip
```
Path Mode: `Full`
```
projects\zip
```
Path Mode: `Absolute`
```
C_\projects\zip
```

Additionally with this being the default behavior, I can't collect files at different paths and add them the root of the zip:
```
'C:\dir\root.txt', 'C:\dir\subdir\sub.txt' | Add-ZipArchiveEntry -ZipArchivePath 'archive.zip' -BasePath 'C:\dir'
```
Results in:
```
root.txt
subdir\sub.txt
```
Instead of:
```
root.txt
sub.txt
```

Reply:
I could go with the default behavior being adding relatively. If you pass in a file, then it gets added to the root of the package. If you pass in a directory, it also gets added to the root.

I don't think we need a PathMode. You can currently use `ParentEntryPath` to control an item's parent path in the archive. Given path `C:\projects\zip`, by default it would get added as zip, but the user could pass `C_\projects` or `projects` to get the `Full` and `Absolute` behavior.

Reply:
Yeah that's a simple enough solution.

A `PathMode` parameter would be additional functionality of making `EntryParentPath` be inferred automatically when used with a `Full` or `Absolute` path mode. Currently, if two items passed to the function have differing parent paths, you would need to call `Add-ZipArchiveEntry` twice with two different values for `EntryParentPath`.

For example:
```
'C:\projects\zip' | Add-ZipArchiveEntry -EntryParentPath 'projects'
'C:\other\misc\configs' | Add-ZipArchiveEntry -EntryParentPath 'other\misc'
```
The same folder structure in the archive could be inferred dynamically:
```
'C:\projects\zip', 'C:\other\misc\configs' | Add-ZipArchiveEntry -PathMode 'Full'
```